packages/flutter/lib/src/material/chip.dart

```diff
@@ -103,81 +111,658 @@ class Chip extends StatelessWidget {
+  _ChipRenderTheme get theme => _theme;
+  _ChipRenderTheme _theme;
+  set theme(_ChipRenderTheme value) {
+    if (_theme == value) {
+      return;
+    }
+    _theme = value;
+    markNeedsLayout();
+  }
+
+  TextDirection get textDirection => _textDirection;
+  TextDirection _textDirection;
+  set textDirection(TextDirection value) {
+    if (_textDirection == value) {
+      return;
+    }
+    _textDirection = value;
+    markNeedsLayout();
+  }
+
+  // The returned list is ordered for hit testing.
+  Iterable<RenderBox> get _children sync* {
     if (avatar != null) {
-      startPadding = 0.0;
-      children.add(new ExcludeSemantics(
-        child: new Container(
-          margin: const EdgeInsetsDirectional.only(end: 8.0),
-          width: _chipHeight,
-          height: _chipHeight,
-          child: avatar,
-        ),
-      ));
+      yield avatar;
     }
+    if (label != null) {
+      yield label;
+    }
+    if (deleteIcon != null) {
+      yield deleteIcon;
+    }
+    if (container != null) {
+      yield container;
+    }
+  }
 
-    children.add(new Flexible(
-      child: new DefaultTextStyle(
-        overflow: TextOverflow.ellipsis,
-        style: labelStyle,
-        child: label,
-      ),
-    ));
-
-    if (deletable) {
-      endPadding = 0.0;
-      children.add(new GestureDetector(
-        onTap: Feedback.wrapForTap(onDeleted, context),
-        child: new Tooltip(
-          message: deleteButtonTooltipMessage ?? MaterialLocalizations.of(context).deleteButtonTooltip,
-          child: new Container(
-            padding: const EdgeInsets.symmetric(horizontal: 4.0),
-            child: new Icon(
-              Icons.cancel,
-              size: 24.0,
-              color: deleteIconColor ?? Colors.black54,
-            ),
-          ),
-        ),
-      ));
+  @override
+  void handleEvent(PointerEvent event, BoxHitTestEntry entry) {
+    assert(debugHandleEvent(event, entry));
+    if (event is PointerDownEvent && deleteIcon != null) {
+      _tap.addPointer(event);
     }
+  }
 
-    return new Semantics(
-      container: true,
-      child: new Container(
-        constraints: const BoxConstraints(minHeight: _chipHeight),
-        padding: new EdgeInsetsDirectional.only(start: startPadding, end: endPadding),
-        decoration: new ShapeDecoration(
-          color: backgroundColor ?? Colors.grey.shade300,
-          shape: border,
-        ),
-        child: new Center(
-          widthFactor: 1.0,
-          heightFactor: 1.0,
-          child: new Row(
-            children: children,
-            mainAxisSize: MainAxisSize.min,
-          ),
-        ),
-      ),
+  void _handleTapDown(TapDownDetails details) {
+    if (deleteIcon != null) {
+      _tapDownLocation = globalToLocal(details.globalPosition);
+    }
+  }
+
+  void _handleTap() {
+    if (_tapDownLocation == null) {
+      return;
+    }
+    if (deleteIcon != null && _deleteButtonRegion.contains(_tapDownLocation)) {
+      onDeleted();
```

Comment:
Is there a corner case where onDeleted is null? Maybe the widget is rebuilt with onDeleted:null in between the press and tap events?

Reply:
Good catch. Yeah, unlikely but possible.